Dev 1.0 - 050116

jmraymundo opened the pull request:
**Concept:**

An Android application that is used for basketball statistics tracking in a live basketball game

**Functions included in the development:**
- Track New Game
  - Add Players/Teams
  - Track Player/Team statistics via user interaction
  - Automatically update specific statistics without user interaction (example: playing time of a player ingame should update while game is ongoing. playing time should not increase when game is stopped.)
- View statistics of previously tracked games
  - View team statistics of previously tracked games

Commit:
Added new android project

Commit:
Transfered Activity to correct package

Commit:
Created empty model objects to be used for data storage and handling.

Commit:
Format update

Commit:
Updated models. Updated activity with initial layout

Commit:
Updated models

Commit:
Added JUnit test cases

Commit:
Updated Team. Now using SparseArray for Player container.

Commit:
Added GameEvents.

Commit:
Updated GameEvents. Will now resolve a GameEvent, and all appended
GameEvents.

Commit:
Updated TestCase for Android JUnit.

Commit:
Updated Game model. Now stores GameEvents using Game Clock and Periods
as keys.

Commit:
Added JSON object conversion

Commit:
Updated models for JSON conversion. Also updated unit tests.

Commit:
Added JSON Serializer and GameEvent Deserializer. Moved all appropriate methods.

Commit:
Updated model. Added ID for Games for easier tracking. Added new handler
for saving and loading the games.

Commit:
Added Date and ID as attributes of a Game

Commit:
Added activities for view and model handling.

Commit:
Added layout for New Team.

Commit:
Added new handling of adding of players into teams. Not yet complete!

Commit:
Added the checks required for New Team creation

Commit:
setup layout for fragment population

Commit:
Updated manifest to fix issue with orientation of game screen.

Commit:
Split up the Game layout into multiple fragments

Commit:
Added new string for Game Log

Commit:
Added new field that stores the actual game time the user starts a new
event. This will be used in event logging, instead of actual game time
the user ends the input of the event.

Commit:
New fragment that displays the names and scores of both teams, game
clock and shot clock, and the current period.

Commit:
Removed custom exception. Altered structure of shooting fouls.

Commit:
Fixed orientation of other activities to be in portrait.

Commit:
Added strings for View.

Commit:
Added handling of Team names.

Commit:
Added team fouls per period. Also added time field for OT situations.

Commit:
Extracted listener to utils package.

Commit:
New dialog for handling queries for setting of game/shot clocks.

Commit:
Added handling of getting the total points scored by a player.

Commit:
Modified tests based on changes made to the Game.

Commit:
New fragment for handling of displaying the currently in-game players.

Commit:
Started implementation of the GameActivity.

Commit:
Fixed confirmation dialog for adding of clock values

Commit:
Overridden toString method for better aesthetics.

Commit:
Added new string resources for dialog for team starters

Commit:
Added finish method to terminate activity since the user should not be
able to go back.

Commit:
Implemented initialization of UI

Commit:
Updated layout for proper positioning.

Commit:
Added new strings. Updated layout.

Commit:
Added null handling in updateUI method.

Commit:
Added setting of ball image to team with ball possession.

Commit:
Updated clock checking. Updated adding of new event. Can now only add
parent events. Child events are appended so no problems with that.

Commit:
Added checking if GameEvent is a child of another.

Commit:
Update button ids and texts. Also update width to use 0dp.

Commit:
Added new strings

Commit:
Removed unused method/field

Commit:
Added new dialog layout for Time-related dialog.

Commit:
Add dialog layout container for simple dialogs with a single textview.

Commit:
Added new basketball image to be used for displaying current team with
ball possession.

Commit:
Removed unused method/field.

Commit:
Updated test case to reflect the correct method name.

Commit:
Added getter for team without ball possession.

Commit:
Added multiple button listeners. 

setup ResetShotClock
setup ResetGameClock
done with Pause/Resume game
done with TurnoverEvents and all branches.

Commit:
added checking of turnover events. Ball possession will now transfer properly.

Commit:
Refactored methods. Turned private all methods not used outside the
activity, if any.

Commit:
Multiple changes

Updated class name from GameMainStatsFragment to GameScoreBoardFragment
Moved some utility methods into a single utility class called
StringUtils
Added container and fragment for handling of GameLog UI.

Commit:
Updated README

Commit:
Added new strings. Updated some strings to reflect usage.

Commit:
Updated layout to reflect new string name

Commit:
Added new dialog layout for Coach button

Commit:
Multiple changes in GameActivity.

Remove unused method.
Updated string names.
Started implementation of Coach button, and relevant effects.

Commit:
Multiple string and fields renames.

Commit:
Updated GameLog string handling and output.

Commit:
Improvements on GameActivity.

Add new method for handling new game events.
Added functionality for resetting of game clock.

Commit:
Updated GameLog UI

Now uses ListView as view.
Updated Strings displayed per period.

Commit:
Updated strings for game/shot clock reset

Commit:
Added string representation of a date

Uses the format YYMMDD.
Updated Game class to reflect and use new string representation to
include the date in the title.

Commit:
Improved GameActivity

Updated method names for readability.
Added checking for fragment visibility before updating UI.
Added functionality for shot clock resetting.
Now properly terminates the timer when the activity is cancelled.
Updated some references to updated strings.

Commit:
Improvements in clock pausing scenarios

Game now pauses when clocks are initialized.
Clocks are reinitialized when going to next period.

Commit:
New implementation: Next period

Added new button for going to next period.
Added strings for next period button.
Added functionality for going to next period.

cristinal commented:
error occurred after saving the progress game

01-05 14:00:03.611: E/AndroidRuntime(23292): java.lang.IllegalStateException: Fragment GameScoreBoardFragment{31a78071} not attached to Activity
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at android.app.Fragment.getResources(Fragment.java:788)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at com.example.bballstatstrack.fragments.GameScoreBoardFragment.getResizedBasketball(GameScoreBoardFragment.java:97)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at com.example.bballstatstrack.fragments.GameScoreBoardFragment.setBallPossession(GameScoreBoardFragment.java:79)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at com.example.bballstatstrack.fragments.GameScoreBoardFragment.updateUI(GameScoreBoardFragment.java:59)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at com.example.bballstatstrack.activities.GameActivity.updateScoreBoardUI(GameActivity.java:537)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at com.example.bballstatstrack.activities.GameActivity.access$7(GameActivity.java:535)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at com.example.bballstatstrack.activities.GameActivity$UpdateTimeTask$1.run(GameActivity.java:719)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at android.os.Handler.handleCallback(Handler.java:739)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at android.os.Handler.dispatchMessage(Handler.java:95)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at android.os.Looper.loop(Looper.java:135)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at android.app.ActivityThread.main(ActivityThread.java:5254)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at java.lang.reflect.Method.invoke(Native Method)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at java.lang.reflect.Method.invoke(Method.java:372)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
01-05 14:00:03.611: E/AndroidRuntime(23292): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)

cristinal commented:
button for turnover of stolen ball is not working

Commit:
Removed extra return.

Commit:
Added new implementation:

Added functionality for FoulButton.
Added initial functionality for free throws.
Added handling of NonShootingFoul events.

Commit:
Fixed issue with nextPeriodButton enabled when gameClock is not 0.

Commit:
Added string for penalty notification.

Commit:
Some minor corrections

Removed append override since it's not needed.
Updated toString since it doesn't append anything.

Commit:
Updated ShootingFoulEvent

Modified constructor. Now needs number of FTs to shoot.
Removed append override. Not needed anymore.
Modified toString to reflect changes in constructor.
Added getter for new field variable.
Modified resolve to reflect changes.

Commit:
Updated affected codes by changes in ShootingFoulEvent.

Removed unused imports and added new field identifier in
ShootingFoulEvent.
Updated JSONSerializer to add new field in JSONObject.
Updated GameEventDeserializer to handle new field from JSONObject.
Updated TestGameEventTOJson to reflect changes in ShootingFoulEvent.

Commit:
Added new String for FreeThrow.

Commit:
Updated test case

cristinal commented on BBallStatsTrack/src/com/example/bballstatstrack/adapter/GameAdapter.java:
```diff
@@ -0,0 +1,43 @@
+
+        return convertView;
+    }
+}
```
please use ViewHolder as a good practice
http://developer.android.com/training/improving-layouts/smooth-scrolling.html#ViewHolder

cristinal replied:
checked ok

Commit:
Added additional handling for SubstitutionEvent

Also added changes to the GameAdapter. Now utilizes ViewHolder.

Commit:
Sorted methods by visibility.

Commit:
Multiple implementation regarding Substitution and Shoot.

Implemented Substitution functionality.
Implemented initial dialog for Shoot functionality.
Added strings to be used for Shoot.
Modified some string names for better readability.
Added new layout for Substitution Button dialog.

Commit:
New strings for ShootEvent dialogs.

Commit:
New layout for ShotType dialog

Commit:
Layout renaming

Commit:
Added new check for ShootEvents.

Possession will now change accordingly.
ShotClock would also reset to correct time.

Commit:
New layout for Dialog of missed shots handling.

Commit:
Various additions.

Refactoring of some dialog creation. Some have similar structures.
Complete implementation of SubstitutionEvent.
Initial implementation of ShootEvent.

Commit:
Updated handling of change of possession.

Shot clock will now always reset to 24 seconds when change of ball
possession occurs.

Commit:
Removed unused strings and layout

Commit:
Game now pauses when the ball went out of bounds.

Commit:
Added implementation for AssistEvents and ReboundEvents.

Commit:
Autoformat change

Commit:
Modified checking if shot clock and game clock are still valid.

Commit:
Added support for Save Game dialog.

Commit:
Updated resetting the shot clock to max.

Takes into consideration the time elapsed due to some events not pausing
the game.

Commit:
Fixed issues with storing and fetching data into JSONObjects.

Commit:
Implemented handling of Save Game.

User can now save the game state whenever a period ends.

Commit:
Updated GameAdapter to utilize ViewHolder.

Commit:
Refactored some methods for proper order.

Commit:
Added new Activity and Fragment for viewing of old games tracked.

Commit:
Added check for substitution.

User cannot substitute players if the team has no bench players.

Commit:
Set some dialogs be uncancelable.

Commit:
Added new method in StringUtils

Converts a time in String format MM:SS to int totalSeconds.

Commit:
New layout for GameReviewActivity

Commit:
Added new methods.

get2ptFGMade() - gets the sum of get2ptFGMade() of all Players in team
get2ptFGMiss() - gets the sum of get2ptFGMiss() of all Players in team
get3ptFGMade() - gets the sum of get3ptFGMade() of all Players in team
get3ptFGMiss() - gets the sum of get3ptFGMiss() of all Players in team
getFTMade() - gets the sum of getFTMade() of all Players in team
getFTMiss() - gets the sum of getFTMiss() of all Players in team

Commit:
Added some changes to Game and Team

Game:
	Added support for updating team possession time.
	
Team:
	Added support for updating team possession time.
	Added getter methods for some player-based stats.

Commit:
Added new utility methods.

getShotFraction - returns a fraction with the format shotMade/shotTotal
getShotPercentage - returns the percentage (###.##%) based on the
formula shotMade/shotTotal * 100

Commit:
Sorted the methods by rearranging based on visibility.

Commit:
StringUtils-related changes

Updated getShotPercentage to use double instead of int.
Update TestStringUtils. Added tests for getShotFraction and
getShotPercentage.

Commit:
Converted UUID to String before putting in intent

Commit:
Implemented GameReview. Now able to view previously tracked games.

Added new layouts for the TeamReviewFragment and GameReviewActivity.
Added strings for the view.
Added GameReviewActivity to the manifest.

Commit:
Implemented new way of writing/reading JSON files.

Now uses JsonReader and JsonWriter.
Removed support of old format (initializing via JSONObject/JSONArray)

cristinal commented on BBallStatsTrack/src/com/example/bballstatstrack/models/Game.java:
```diff
@@ -0,0 +1,496 @@
+
+    private final int mMaxGameClock;
+
+    private final int mMaxOTGameClock;
```
Are you still going to use mMaxOTGameClock?

jmraymundo replied:
Yes. Game clock would reset to this value if mPeriod > 4. Currently, it is set to 5, since this is by FIBA standards. In the future, user would also be asked to set this value, if they wish.

cristinal replied:
ok :+1: . file updated and confirmed that mMaxOTGameClock is being used.

cristinal commented on BBallStatsTrack/src/com/example/bballstatstrack/models/Game.java:
```diff
@@ -0,0 +1,496 @@
+
+    private Team mHomeTeam;
+
+    private Team mHasBallPossession;
```
I think there is other way to save data as to whose team is in possession of the ball. Instead of using Team object, we could use id as comparison or simple boolean condition like mHomeHasBallPossession... something like that.

jmraymundo replied:
mHasBallPossession is used specifically as a reference to both mAwayTeam or mHomeTeam. I implemented it in Game instead of inside of Team because I think the Team object doesn't really care if it has ball possession or not. The Game decides who has the ball. I also implemented it this way so that it would be impossible to have both teams being in ball possession. However, if you have more ideas, please feel free to share them.

cristinal replied:
Agrees with Game should handle ball possession but before mHasBallPossession was used to only check which team has the ball and swap and I think that using Team object would make the code look complicated. 

Now with the updated file, it is used to update time. This is ok.

cristinal commented on BBallStatsTrack/src/com/example/bballstatstrack/activities/AddPlayersToTeamsActivity.java:
```diff
@@ -0,0 +1,205 @@
+
+    protected static final String AWAY_TEAM_NUMBERS = "AwayTeamNumbers";
+
+    private String mHomeName;
```
please add private access modifier if variable is used only here

jmraymundo replied:
Done

cristinal replied:
:+1:

cristinal commented on BBallStatsTrack/src/com/example/bballstatstrack/activities/AddPlayersToTeamsActivity.java:
```diff
@@ -0,0 +1,205 @@
+        {
+            mHomeButton.setEnabled( true );
+            mAwayButton.setEnabled( true );
+        }
```
```
else
        {
            mHomeButton.setEnabled( true );
            mAwayButton.setEnabled( true );
        }
```

not used?

jmraymundo replied:
Removed

cristinal replied:
:+1:

cristinal commented on BBallStatsTrack/res/layout/dialog_new_player.xml:
```diff
@@ -0,0 +1,35 @@
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+	android:layout_width="match_parent"
+	android:layout_height="match_parent"
+	android:orientation="vertical" >
```
please kindly check.... LinearLayout is possibly useless?

jmraymundo replied:
Removed LinearLayout. It was indeed useless.

cristinal replied:
:+1:

Commit:
Major changes to GameActivity.

Added support for ShootingFoulEvent.
Modified the code related to ShootEvents due to interaction with
ShootingFoulEvents.
Modified tracking of time of GameEvents. Now part of GameEvent.
Previously acts as key for storing GameEvents in SparseArray. Due this
this change, PeriodLogs now a List, instead of a SparseArray. GameLog
changes are made to reflect changes to the PeriodLogs.
Fixed issue with overwriting events during stop time.
Changing of ball possession now acting correctly. Previously had a weird
interaction with ShootingFoulEvents due to possible multiple ShootEvents
simultaneously.

Commit:
Import cleanup and method sorting.

cristinal commented on BBallStatsTrack/src/com/example/bballstatstrack/activities/AddPlayersToTeamsActivity.java:
```diff
@@ -0,0 +1,205 @@
+
+    private class ProceedButtonListener implements OnClickListener
+    {
+        public boolean hasCompleteTeamNames()
```
At first read, "hasCompleteTeamNames" name could indicate that both teams have submitted their names. please rename the method. thanks.

jmraymundo replied:
Done

cristinal replied:
checked ok

Commit:
PR issue fix

dialog_new_player.xml:
Remove unnecessary LinearLayout

AddPlayersToTeamsActivity.java:
Renamed method hasCompleteTeamNames() to hasIncompleteTeamNames()
Added private access modifier to various fields

cristinal commented on BBallStatsTrack/src/com/example/bballstatstrack/activities/AddPlayersToTeamsActivity.java:
```diff
@@ -0,0 +1,205 @@
+            setIntentExtras( intent, mHomeTeam, HOME_TEAM_NUMBERS, HOME_TEAM_MEMBER_NAMES );
+            intent.putExtra( AWAY_TEAM_NAME, mAwayName );
+            setIntentExtras( intent, mAwayTeam, AWAY_TEAM_NUMBERS, AWAY_TEAM_MEMBER_NAMES );
+            finish();
```
instead of adding too many extras, why not add Team object as extra?

cristinal commented on BBallStatsTrack/src/com/example/bballstatstrack/activities/AddPlayersToTeamsActivity.java:
```diff
@@ -0,0 +1,205 @@
+            intent.putExtra( HOME_TEAM_NAME, mHomeName );
+            setIntentExtras( intent, mHomeTeam, HOME_TEAM_NUMBERS, HOME_TEAM_MEMBER_NAMES );
+            intent.putExtra( AWAY_TEAM_NAME, mAwayName );
+            setIntentExtras( intent, mAwayTeam, AWAY_TEAM_NUMBERS, AWAY_TEAM_MEMBER_NAMES );
```
instead of adding too many extras, why not add Team object as extra?

jmraymundo replied:
mHomeTeam/mAwayTeam, at this point, are of type SparseArray<String>. The keys are used as player numbers, and the values are the players' names. The intent is passed on to GameActivity, which creates the Team objects.

cristinal replied:
how about creating Team objects here instead in the GameActivity?
would there be a difference?

Commit:
Updated showBallPossessionDeciderDialog() to be uncancellable.

cristinal commented:
GameActivity.java:
this line of code should not be included in setClockFromDialog
```clockField.addTextChangedListener( new PlayerNumberWatcher( 0, max ) );```
it does not filter the text during the set up and only functions after clicking ok.

cristinal commented on BBallStatsTrack/src/com/example/bballstatstrack/models/Game.java:
```diff
@@ -0,0 +1,434 @@
+        }
+    }
+
+    private void checkTeamReboundEvent( GameEvent event )
```
method not in use. please kindly remove

jmraymundo replied:
Done

cristinal replied:
:+1:

cristinal commented on BBallStatsTrack/src/com/example/bballstatstrack/models/Game.java:
```diff
@@ -0,0 +1,434 @@
+        mPeriod = mGameLog.getCurrentPeriod();
+    }
+
+    private boolean isShotMade( GameEvent event )
```
method not in use. please kindly remove

jmraymundo replied:
Done

cristinal replied:
:+1:

Commit:
Removed unnecessary final access modifier

Commit:
PR issue fixes

GameActivity.java
	Updated inflation of dialog layouts.
	Reworked setting of TextWatcher to the clock-related dialogs.
	
Game.java
	Removed some unused methods.

raymundoklab commented:
Regarding this issue

> GameActivity.java:
> this line of code should not be included in setClockFromDialog
> clockField.addTextChangedListener( new PlayerNumberWatcher( 0, max ) );
> it does not filter the text during the set up and only functions after clicking ok.

Issue is now fixed. Reworked this. Instead of passing the dialog, the EditText is now passed and evaluated. Setting of the listener is now part of the declaration, as it should be in the first place.

raymundoklab commented:
I am now currently updating the handling of the Timer. As you suggested, I am on planning on stopping the timer when the game is paused, and restarting it when it's ongoing.

Commit:
Dismiss the dialog after selecting

cristinal commented:
1. saves log
2. exit app
3. return to app
4. error occurs while attempting to display log list
```
01-11 10:14:29.350: E/AndroidRuntime(10963): Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was NAME
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at android.util.JsonReader.expect(JsonReader.java:310)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at android.util.JsonReader.beginObject(JsonReader.java:293)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at com.example.bballstatstrack.models.utils.JSONDeserializer.readEvent(JSONDeserializer.java:130)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at com.example.bballstatstrack.models.utils.JSONDeserializer.readPeriodLog(JSONDeserializer.java:264)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at com.example.bballstatstrack.models.utils.JSONDeserializer.readGameLog(JSONDeserializer.java:238)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at com.example.bballstatstrack.models.utils.JSONDeserializer.readGame(JSONDeserializer.java:72)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at com.example.bballstatstrack.models.utils.JSONFileSaverLoader.loadGames(JSONFileSaverLoader.java:40)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at com.example.bballstatstrack.model.GameDirectory.<init>(GameDirectory.java:32)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at com.example.bballstatstrack.model.GameDirectory.get(GameDirectory.java:50)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at com.example.bballstatstrack.fragments.ListGamesFragment.onCreate(ListGamesFragment.java:56)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at android.app.Fragment.performCreate(Fragment.java:2031)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:863)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1067)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at android.app.BackStackRecord.run(BackStackRecord.java:834)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1452)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at android.app.Activity.performStart(Activity.java:6005)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2288)
01-11 10:14:29.350: E/AndroidRuntime(10963): 	... 10 more

```

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/activities/AddPlayersToTeamsActivity.java:
```diff
@@ -0,0 +1,198 @@
+                names[i] = team.valueAt( i );
+            }
+            intent.putExtra( extraTeamNumbersIdentifier, numbers );
+            intent.putExtra( extraTeamNamesIdentifier, names );
```
Creating a team object which will hold List of Players as object(playerNumber, steal, points, rebouns and etc) will also simplify this, it will also be a good documentation.

raymundoklab commented:
As per both of your suggestions, I am refactoring the Team creation part. Team objects are going to be Parcelable.

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/adapter/GameAdapter.java:
```diff
@@ -0,0 +1,52 @@
+
+public class GameAdapter extends ArrayAdapter< Game >
+{
+    Activity mActivity;
```
please add access modifier if this is used only here po

jmraymundo replied:
Variable mActivity removed since it's not needed anymore.

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/adapter/GameAdapter.java:
```diff
@@ -0,0 +1,52 @@
+    {
+        if( null == convertView )
+        {
+            convertView = mActivity.getLayoutInflater().inflate( R.layout.list_games, null );
```
Just to note po you can also use
LayoutInflater.from(parent.getContext());
To get the inflater instead of asking for an activity in the constructor just for the sake of obtaining the LayoutInflater po :bow:

jmraymundo replied:
Done

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/fragments/TeamFragment.java:
```diff
@@ -0,0 +1,308 @@
+            return;
+        }
+        int playerNumber = Integer.parseInt( playerNumberString );
+        String playerName = playerNameString;
```
Probably not necessary to do this po. :bow:

jmraymundo replied:
Functionality transferred to NewPlayerDialogListener. Done!

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/fragments/TeamFragment.java:
```diff
@@ -0,0 +1,308 @@
+        dialog.getWindow().setSoftInputMode( WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE );
+        dialog.show();
+        dialog.setCanceledOnTouchOutside( false );
+        setPositiveButtonAction( dialog );
```
You can remove this line of code po and instead move the addNewPlayerFromDialog method in the builder.setPositiveButton( android.R.string.ok, new DialogInterface.OnClickListener() po.

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/Game.java:
```diff
@@ -0,0 +1,406 @@
+    {
+        if( team.equals( mAwayTeam ) )
+        {
+            mHasBallPossession = mAwayTeam;
```
I think it will be ok to just do
mHasBallPossession = team
since your condition is team.equals po .. this will eliminate the if else conditions

jmraymundo replied:
Since it is impossible to set a team that is neither the mHomeTeam and mAwayTeam, this is updated.

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/model/GameDirectory.java:
```diff
@@ -0,0 +1,96 @@
+        saveGames();
+    }
+
+    public Game getCrime( UUID id )
```
Please change method name po.

jmraymundo replied:
Oops my bad! Done!

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/gameevents/TurnoverEvent.java:
```diff
@@ -0,0 +1,69 @@
+        if( appendedEvent instanceof StealEvent && mTurnoverType == TurnoverType.STEAL )
+        {
+            mAppended = appendedEvent;
+            return;
```
Probably not needed po since the remaining lines beneath does nothing once the if condition for this code block is true.

jmraymundo replied:
done

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/gameevents/TurnoverEvent.java:
```diff
@@ -0,0 +1,69 @@
+                return;
+            }
+        }
+        super.append( appendedEvent );
```
can be removed since GameEvent.append(GameEvent appendedEvent) does nothing po..

jmraymundo replied:
done

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/Game.java:
```diff
@@ -0,0 +1,406 @@
+        ID( "id" ), AWAY_TEAM( "awayTeam" ), HOME_TEAM( "homeTeam" ), GAME_LOG( "gameLog" ), PERIOD_LOG(
+                "periodLog" ), DATE( "date" );
+
+        private final String mConstant;
```
You can instead make this a public variable, since this is already declared as final there will be no confusion or accident in which the value might change.

jmraymundo replied:
As practice, I do not allow fields to be accessed outside of the scope, and instead use a getter method. I have updated this to use a getter method instead of overriding toString().

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/Game.java:
```diff
@@ -0,0 +1,406 @@
+        }
+
+        @Override
+        public String toString()
```
If you will change the access modifier of mConstant you can remove this and instead directly call GameStats.mConstant (e.g. ID.mConstant)

jmraymundo replied:
As practice, I do not allow fields to be accessed outside of the scope, and instead use a getter method. I have updated this to use a getter method instead of overriding toString().

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/utils/JSONDeserializer.java:
```diff
@@ -0,0 +1,431 @@
+        {
+            return mHomeTeam;
+        }
+        else
```
remove else block and instead just return null po

jmraymundo replied:
done

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/utils/JSONDeserializer.java:
```diff
@@ -0,0 +1,431 @@
+        {
+            return mAwayTeam;
+        }
+        else
```
remove else block and instead just return null po

jmraymundo replied:
done

cristinal commented on BBallStatsTrack/res/layout/fragment_team_ingame_stats.xml:
```diff
@@ -0,0 +1,200 @@
+            android:layout_weight="1"
+            android:background="@drawable/cell_border"
+            android:gravity="center"
+            android:textSize="10sp" />
```
recommending to use dimen.xml
for layout_width, layout_height, textSize, paddings and margins

jmraymundo replied:
This is noted. This will be done in the future.

_update 150116_ Done!

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/Game.java:
```diff
@@ -0,0 +1,406 @@
+
+import android.text.format.DateFormat;
+
+public class Game
```
Game can actually be implemented as a Singleton since there will always be a single instance of this on the app. Making this a singleton will let you lessen the times Game is passed in the parameter and instead use Game.getInstance() if an instance of Game is needed.

Sample : https://sourcemaking.com/design_patterns/singleton/java/1

jmraymundo replied:
There can be multiple instances of a Game. After a game has been recorded, it is stored in the memory. This can be access later and the final state reviewed by the user,

lelelelem replied:
Ahh I see Game is actually used also in the list. Ok Noted :+1:

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/Player.java:
```diff
@@ -0,0 +1,265 @@
+        setFullName( fullName );
+    }
+
+    public Player( int number, String fullName, int miss1pt, int miss2pt, int miss3pt, int made1pt, int made2pt,
```
Builder patterns are best for situations such as this(Constructors having too many parameters). If you still have time you can implement it here.

lelelelem replied:
Builder patterns are best for situations such as this(Constructors having too many parameters, some of which not required). If you still have time please implement it here.

lelelelem replied:
[Builder Pattern Sample 1](http://www.javaworld.com/article/2074938/core-java/too-many-parameters-in-java-methods-part-3-builder-pattern.html)

[Effective Java (page 11) Builder Pattern](https://vk.com/doc11181294_320575074?hash=858b952caa4eae4453&dl=3feeed6b6cf08581ce)

jmraymundo replied:
This constructor is special in the case that all parameters are required. Also, when this constructor is called, technically the Player object is in the final state, as no other values will be changed. This is only used when loading the game data and reviewing all of the games stats.

lelelelem replied:
Basically as stats all fields are required but not all parameters needed to be initialized or passed in the constructor, for example

```
new Player( 12, "Player One", 0, 0, 3, 0, 0, 1, 0, 1, 0, 0, 1, 0, 3, 130);
```

Many stats are basically just zero and therefore can have basic values of zero, using a builder pattern will essentially do this by making unset values to zero.
___Example__
![image](https://cloud.githubusercontent.com/assets/7711811/12316895/c2ffb3e4-bac6-11e5-8e49-81866d58e0cd.png)
You can see that fields/parameters which has the potential to have a 0 value is not included in the constructor but instead has an optional setter method and only fields/parameters that needs/required to have a value is asked on the parameter.

lelelelem replied:
But still seeing the changes of your code I wont force you to change your code. But knowing builder pattern will be very helpful in the future po. :bow:

cristinal commented:
As suggestion/tip for future developments, as much as possible avoid having long classes if we could separate other methods as another element, the better...

for example in some of your models, you have included logic part. In my opinion, models (most of the time) are just data holders. Separate logic as util/controller/service part. :smile:

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/utils/JSONDeserializer.java:
```diff
@@ -0,0 +1,431 @@
+            name = reader.nextName();
+            if( name.equals( PlayerStats.NUMBER.toString() ) )
+            {
+                number = reader.nextInt();
```
If you can implement Builder Pattern you can change this to Object.BuilderMethod().setName(String)(you can use the  other setter methods in the builder to replace declarations within the loop), you can also remove the multiple temporary variables declared above.

jmraymundo replied:
Player fields (besides the number and fullName) don't have setter methods. The only time they change value is via constructor, or via its methods that increment them by 1. It is designed this way to keep the fact that a player's stats doesn't change other than incrementing by 1 (example: gain 2 rebounds only by getting 1 rebound twice) I have updated the code the removes the need of the while loop and also the temporary initialization of the variables. They are now initialized to the correct value properly

lelelelem replied:
My comment applies if you will implement the Builder Pattern In my other comment. Implementing a Builder pattern to your Player will also have you create setters for fields inside the Class (as seen on my example). This builder pattern will only be used for building a Player instance from a JSON file.

lelelelem replied:
No need to change but it will be good to be familiar with the builder patter po :bow:

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/utils/JSONDeserializer.java:
```diff
@@ -0,0 +1,431 @@
+            }
+        }
+        reader.endObject();
+        return new Player( number, fullName, miss1pt, miss2pt, miss3pt, made1pt, made2pt, made3pt, offReb, defReb,
```
If you can implement Builder Pattern you can change this to Object.BuilderMethod().build() which will pass the needed object.

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/utils/JSONDeserializer.java:
```diff
@@ -0,0 +1,431 @@
+        {
+            type = EventType.valueOf( reader.nextString() );
+        }
+        if( reader.nextName().equals( GameEvent.PLAYER_NUMBER ) )
```
I think it will be better to use else if for the succeeding conditions, since even if the if condition above them is true the other if conditions will still be checked, as opposed to using if else in which once a condition is deemed true other conditions will be skipped.

jmraymundo replied:
Reworked this as well. I removed the  if else conditions and instead just consumed all names since it has an equivalent in the JSONSerializer.

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/Player.java:
```diff
@@ -0,0 +1,265 @@
+                                "assist" ), TURNOVER( "turnover" ), STEAL( "steal" ), BLOCK( "block" ), FOUL(
+                                        "foul" ), PLAYING_TIME( "playingTimeSec" );
+
+        private final String mConstant;
```
You can change this to public since its already final then use PlayerStats.mConstant po instead of overriding toString()

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/gameevents/GameEvent.java:
```diff
@@ -0,0 +1,114 @@
+
+    public enum ShotClass
+    {
+        FT, FG_2PT, FG_3PT;
```
You can add a String parameter to each enum value here so that you can easily maintain and change the messages for each event/enum value

```
FT("Freethrow shot"),...
```

instead of defining them by using switches such as this.

```
switch( mShotClass )
   {
      case FT:
      shotClass = "FT shot. ";
      break;
      //Code Continuation 
    }
```

This comment applies to other of your enum that has message accompanying them.

jmraymundo replied:
Done

Commit:
Done with refactoring

Extracted all Dialogs and the respective Listeners to its own class.
Game, Team and Player are now parcelable.
New object ListPlayer, which simply hold a list of Players, while offering some useful methods.
Some fragments are now Observers, which observes the Game (now Observable)
Updating of game/clock timers reworked.

Commit:
Sorted some methods and organized the imports.

Commit:
Removed unnecessary return.

Commit:
Fixed issue with TurnoverEvents and change of possession.

Commit:
Various changes

Updated manifest. Now enables the app to flip in landscape mode.
Updated GameActivity. Timer is now stopped properly on finish.
Minor update in GameLogFragment. Very small change to a field name.
Updated GameLog. Now correctly displaying logs with the oldest at the
very bottom. Was previously displaying with the oldest at the bottom,
but with the period order with the oldest period at the top.

Commit:
Fixed issue in FreeThrow events.

Now properly showing ReboundDialog when the final free throw misses.

Commit:
Extracted GameActivity button listeners.

Also fixed issue with TurnoverEvents and change of ball possession.
Transferred checking for period fouls to the activity.

Commit:
Various updates.

Replaces the term 'Game Clock' to 'Period Clock' in all derivations..
Updated shot clock. It is now "turned-off" when it's less than the
period clock.
Updated SubstitutionEvent string.

lelelelem commented on BBallStatsTrack/src/com/example/bballstatstrack/models/gameevents/GameEvent.java:
```diff
@@ -0,0 +1,122 @@
+
+    public enum ShotClass
+    {
+        FT( "FT shot" ), FG_2PT( "2pt shot" ), FG_3PT( "3pt shot" );
```
You can apply this same format for other of your enums which has corresponding message with them such as Rebound and etc po..

jmraymundo replied:
ReboundType is not applicable since there are 2 kinds of rebounds: Player rebounds and team rebounds. All in all, this makes the total rebound types 4: Offensive/Defensive player rebounds, and Offensive/Defensive team rebounds, which results in different texts to be shown. 
This, however, is applicable in the TurnoverType enum.

lelelelem replied:
ok po :+1:

Commit:
Updated TurnoverType enum.

It now stores the resID of the associated text.
Also transfered the ProceedButtonListener into the listeners package.

Commit:
Various bug fixes.

GameActivity now observing Game instead of UI Fragments.
GameActivity will now call the updateUI methods of the UI Fragments.
Fixed issue with FreeThrow events.
Fixed issue with NonShootingFoul events during penalty situation.
Reseting period clock will now trigger reset of shot clock.
Reseting period clock will now clear the period log.
Fixed issue with MaxPeriodClockDialog closing with empty input.
Fixed issue with MaxShotClockResetDialog closing with empty input.
Fixed issue with Shot Clock not showing anymore when previously hidden.

Commit:
Fixed issue with Pause/Unpause.

Button is now disabled if shotclock/periodclock is 0.

Commit:
Set styles for multiple views/layouts having common properties

Also added auto-capitalization of the Player Names when adding a new one

Commit:
Restored textType textPersonName

Commit:
Removed unnecessary variable

Commit:
Added limit of number to 2 since range is 0-99

Commit:
Various changes

Style for table cell of the ingame players updated. Text size reduced.
AddPlayersToTeamsActivity now hiding the keyboard when touched.
Fixed bug in TeamReviewFragment which displays the steal count in the
assists row.

Commit:
Fixed issue with FoulButtonDialog and FoulTypeDialog cancellation.

Fixed issue with cancelling Foul Button Dialog unpausing the game even
though the pause is not caused by the Foul button. Also fixed issue with
FoulTypeDialog not showing FoulButtonDialog on cancel.

Commit:
Period clock will be set to 0 if it becomes less than 0.

cristinal commented on BBallStatsTrack/res/layout/activity_new_team.xml:
```diff
@@ -0,0 +1,46 @@
+	android:id="@+id/newTeamContainer"
+	android:layout_width="match_parent"
+	android:layout_height="match_parent"
+	android:orientation="vertical" >
```
please remove this orientation

jmraymundo replied:
done

cristinal replied:
:+1:

cristinal commented on BBallStatsTrack/res/layout/activity_new_team.xml:
```diff
@@ -0,0 +1,46 @@
+		android:layout_width="match_parent"
+		android:layout_height="wrap_content"
+		android:layout_alignParentTop="true"
+		android:layout_centerHorizontal="true" >
```
add orientation attribute if it is LinearLayout

jmraymundo replied:
done

cristinal replied:
:+1:

Commit:
Updated orientation of some layouts.

Removed orientation from RelativeLayout.
Added orientation in LinearLayout.
